### ISSUE TITLE
kernel/pgo for raw profile format version 6

### ISSUE THREAD

**Issue by MaskRay** [ClangBuiltLinux/linux#1405]
If https://reviews.llvm.org/D104556 makes it into LLVM 13.0.0 (likely), the following diff will ideally be needed in kees/for-next/clang/pgo

```diff
diff --git a/kernel/pgo/fs.c b/kernel/pgo/fs.c
index 3c5aa7c2a4ce..6a023f95142d 100644
--- a/kernel/pgo/fs.c
+++ b/kernel/pgo/fs.c
@@ -66,7 +66,12 @@ static void prf_fill_header(void **buffer)
        header->counters_size = prf_cnts_count();
        header->padding_bytes_after_counters = 0;
        header->names_size = prf_names_count();
+#if CONFIG_CLANG_VERSION >= 130000
+       header->counters_delta = (u64)((uintptr_t)__llvm_prf_cnts_start -
+                                      (uintptr_t)__llvm_prf_data_start);
+#else
        header->counters_delta = (u64)__llvm_prf_cnts_start;
+#endif
        header->names_delta = (u64)__llvm_prf_names_start;
        header->value_kind_last = LLVM_INSTR_PROF_IPVK_LAST;
 
--- a/kernel/pgo/pgo.h
+++ b/kernel/pgo/pgo.h
@@ -43,7 +43,11 @@
                 (u64)'R' << 8  |       \
                 (u64)129)
 
+#if CONFIG_CLANG_VERSION >= 130000
+#define LLVM_INSTR_PROF_RAW_VERSION            6
+#else
 #define LLVM_INSTR_PROF_RAW_VERSION            5
+#endif
 #define LLVM_INSTR_PROF_DATA_ALIGNMENT         8
 #define LLVM_INSTR_PROF_IPVK_FIRST             0
 #define LLVM_INSTR_PROF_IPVK_LAST              1
@@ -86,7 +90,7 @@ struct llvm_prf_header {
  * struct llvm_prf_data - represents the per-function control structure.
  * @name_ref: the reference to the function's name.
  * @func_hash: the hash value of the function.
- * @counter_ptr: a pointer to the profile counter.
+ * @counter_ptr: the distance from the data start to the counter.
  * @function_ptr: a pointer to the function.
  * @values: the profiling values associated with this function.
  * @num_counters: the number of counters in the function.
@@ -95,7 +99,7 @@ struct llvm_prf_header {
 struct llvm_prf_data {
        const u64 name_ref;
        const u64 func_hash;
-       const void *counter_ptr;
+       const u64 counter_ptr;
        const void *function_ptr;
        void *values;
        const u32 num_counters;
```

Suggestion to `Documentation/dev-tools/pgo.rst`

The raw profile format (`*.profraw*`) is private in llvm-project (between the runtime and llvm-profdata).
The kernel needs to keep its profraw version matching the current clang default. You can find the current raw profile format version with the following command:

```sh
% echo 'int main(){}' > hello.c; clang -fprofile-generate hello.c; LLVM_PROFILE_FILE=default.profraw ./a.out; file default.profraw
default.profraw: LLVM raw profile data, version 5
```

**Comment by nathanchance** [comment]
`CONFIG_PGO_CLANG` will already implicitly require LLVM 13.0.0 because of [this series](https://lore.kernel.org/r/20210621231822.2848305-1-ndesaulniers@google.com/) from [this discussion](https://lore.kernel.org/r/YMTn9yjuemKFLbws@hirez.programming.kicks-ass.net/), which is going to be needed for PGO to get merged into mainline. If there is going to be a breaking change on the LLVM side, it should be handled on the kernel side at the same time and ideally before the feature is merged into mainline so that we are not having to send something to stable.

My personal vote is that the LLVM patch be merged once the kernel patch is merged into Kees's tree and it be shipped to Linus in that state.

@kees @nickdesaulniers @gwelymernans what do you think?

**Comment by nickdesaulniers** [comment]
see also #1074

**Comment by nathanchance** [comment]
As PGO was rejected upstream in its current (#1074), I am going to close this; a new revision will have to incorporate the suggested changes to stay compatible with LLVM 14 as the change has already been merged: https://github.com/llvm/llvm-project/commit/a1532ed27582038e2d9588108ba0fe8237f01844
